srml/contract/src/wasm/prepare.rs

```diff
@@ -229,7 +229,10 @@ impl<'a, Gas: 'a + As<u32> + Clone> ContractModule<'a, Gas> {
 				.get(*type_idx as usize)
 				.ok_or_else(|| "validation: import entry points to a non-existent type")?;
 
-			if !C::can_satisfy(import.field().as_bytes(), func_ty) {
+			// We disallow importing `gas` function here since it is treated as implementation detail.
+			if import.field().as_bytes() == b"gas"
+				|| !C::can_satisfy(import.field().as_bytes(), func_ty)
+			{
 				return Err("module imports a non-existent function");
```

Comment:
For the `gas` function this error message is irritating, as the function exists but it is just forbidden to import it. For forbidden functions, I would propose a better error message.

Reply:
Ha, I actually started with a separate `if` for this check! The reason why I decided to return the same message is because from the standpoint of a user this function doesn't exists and non-available for import. Besides that, if we really removed this function at some point then message would change.

But I'd agree that this is a weak argument, might confuse developrs of this module, and after all I'd consider error messages being unstable anyway, so would happy to change if you are not convienced.

Reply:
Can a developer somehow reference the `gas` function? If yes, I would like to have a better error message. If the developer would need to write its wasm code by hand to import this function, your error message is fine.

Reply:
You mean contract developers? The idea of this PR is to prevent them from accessing this function : )

If you mean the contract module developer then yeah I don't expect them to use this function directly either. It is only emitted by tools like wasm_utils.

Reply:
Ahh okay, perfect.